src/lib.rs

```diff
@@ -67,21 +76,78 @@ impl CipherSuite for DefaultCipherSuite {
     type OprfCs = opaque_ke::Ristretto255;
     type KeGroup = opaque_ke::Ristretto255;
     type KeyExchange = opaque_ke::key_exchange::tripledh::TripleDh;
-    type Ksf = Argon2<'static>;
+    type Ksf = CustomKsf;
 }
 
 #[cfg(feature = "p256")]
 impl CipherSuite for DefaultCipherSuite {
     type OprfCs = p256::NistP256;
     type KeGroup = p256::NistP256;
     type KeyExchange = opaque_ke::key_exchange::tripledh::TripleDh;
-    type Ksf = Argon2<'static>;
+    type Ksf = CustomKsf;
 }
 
 const BASE64: b64::GeneralPurpose = b64::URL_SAFE_NO_PAD;
 
 type JsResult<T> = Result<T, Error>;
 
+#[derive(Default)]
+struct CustomKsf {
+    argon: Argon2<'static>,
+}
+
+impl Ksf for CustomKsf {
+    fn hash<L: ArrayLength<u8>>(
+        &self,
+        input: GenericArray<u8, L>,
+    ) -> Result<GenericArray<u8, L>, InternalError> {
+        let mut output = GenericArray::default();
+        self.argon
+            .hash_password_into(&input, &[0; argon2::RECOMMENDED_SALT_LEN], &mut output)
+            .map_err(|_| InternalError::KsfError)?;
+        Ok(output)
+    }
+}
+
+fn build_argon2_ksf(
+    t_cost: u32,
+    m_cost: u32,
+    parallelism: u32,
+) -> Result<Option<CustomKsf>, Error> {
+    let mut param_builder = ParamsBuilder::default();
+    param_builder.t_cost(t_cost);
+    param_builder.m_cost(m_cost);
+    param_builder.p_cost(parallelism);
+
+    if let Ok(params) = param_builder.build() {
+        let argon = Argon2::new(Algorithm::Argon2id, Version::V0x13, params);
+        return Ok(Some(CustomKsf { argon }));
+    }
+
+    Err(Error::Internal {
+        context: "Invalid keyStretchingFunctionConfig (argon2id) combination",
+        error: InternalError::KsfError,
+    })
+}
+
+fn get_custom_ksf(
+    ksf_config: Option<KeyStretchingFunctionConfig>,
+) -> Result<Option<CustomKsf>, Error> {
+    let config = ksf_config.unwrap_or(KeyStretchingFunctionConfig::Recommended);
+
+    match config {
+        // https://www.ietf.org/archive/id/draft-irtf-cfrg-opaque-17.html#name-configurations
+        // using the recommended parameters for Argon2id except we due 2^21-1 since 2^21 crashes in browsers
+        KeyStretchingFunctionConfig::Recommended => build_argon2_ksf(1, u32::pow(2, 21) - 1, 4),
+        // https://www.rfc-editor.org/rfc/rfc9106.html#name-recommendations
+        KeyStretchingFunctionConfig::MemoryConstrained => build_argon2_ksf(3, u32::pow(2, 16), 4),
```

Comment:
I'd update the link in the comments for the `MemoryConstrained` config to `https://www.rfc-editor.org/rfc/rfc9106.html#section-4-6.2`, as it gives more of an explanation

Reply:
good one, much better! thx